test/distributed/_composable/fsdp/test_fully_shard_compile.py

```diff
@@ -382,51 +403,81 @@ def input_creation_fn():
+            with self._mock_sdpa(fullgraph), self._mock_reinplace_fsdp_all_gather(
+                fullgraph
+            ):
+                self._test_traceable_fsdp(
+                    *self._create_transformer_factory_fns(),
+                    "aot_eager",
+                    fullgraph=fullgraph,
+                )
 
     @skipIfRocm
     @skip_if_lt_x_gpu(2)
     # TODO: native_dropout has worse accuracy after decomp, need to figure out why
     @torch._inductor.config.patch(fallback_random=True)
-    def test_transformer_fullgraph_backend_aot_eager_decomp_partition(self):
-        self._test_traceable_fsdp(
-            *self._create_transformer_factory_fns(),
-            "aot_eager_decomp_partition",
-            fullgraph=True,
-        )
+    def test_transformer_backend_aot_eager_decomp_partition(self):
+        for fullgraph in [True, False]:
+            with self._mock_sdpa(fullgraph):
+                self._test_traceable_fsdp(
+                    *self._create_transformer_factory_fns(),
+                    "aot_eager_decomp_partition",
+                    fullgraph=fullgraph,
+                )
 
     @skipIfRocm
     @unittest.skipIf(not has_triton(), "Inductor+gpu needs triton and recent GPU arch")
     @skip_if_lt_x_gpu(2)
     # TODO: native_dropout causes CUDA IMA error, need to figure out why
     @torch._inductor.config.patch(fallback_random=True)
-    def test_transformer_fullgraph_backend_inductor(self):
-        with mock.patch.object(
-            comms,
-            "reinplace_fsdp_all_gather",
-            functools.partial(
-                self._reinplace_all_gather_with_checks,
-                orig_fn=comms.reinplace_fsdp_all_gather,
-            ),
-        ):
-            _, triton_codes = run_and_get_code(
-                lambda: self._test_traceable_fsdp(
-                    *self._create_transformer_factory_fns(), "inductor", fullgraph=True
+    def test_transformer_backend_inductor(self):
+        for fullgraph in [True, False]:
+            with self._mock_sdpa(fullgraph), self._mock_reinplace_fsdp_all_gather(
+                fullgraph
+            ):
+                _, triton_codes = run_and_get_code(
+                    lambda: self._test_traceable_fsdp(
+                        *self._create_transformer_factory_fns(),
+                        "inductor",
+                        fullgraph=fullgraph,
+                    )
+                )
+            if fullgraph:
+                self.assertTrue(
+                    len(triton_codes) == 2,
+                    "Expected two separate lowerings to Triton code, one from FWD graph and one from Compiled Autograd BWD graph",
+                )
+                for code in triton_codes:
+                    FileCheck().check(
+                        "torch.ops._c10d_functional.all_gather_into_tensor_out."
+                    ).run(code)
+            else:
+                self.assertTrue(
+                    len(triton_codes) >= 3,
```

Comment:
do you think it would be useful to be a bit stricter about what we assert? If I understand properly, my understanding is something like:

3 graphs total (2 fw's due to graph break, 1 unified bw due to compiled autograd)

fw 1: only comm is a `all_gather_out`, and one set_() (performs the weight AG, but does not free the weight)
fw2: no comms, only contains a set_() (just finishes the fw compute and frees the weight)
bw: contains `all_gather_out`, `reduce_scatter_out`, and 2`set_()` ops (fully gathers weight, does bw compute and fress)

(or at least, why is the assert >=3 and not just == 3?)

Reply:
yes I think we want to be more strict here - there is some recompile happening when there is graph break, I'll look into it with @anijain2305 and then add more strict checks